### PR TITLE
No really, fix msg ordering

### PR DESCRIPTION
**Pull request by nicpottier**
3rd time is a charm?

So after doing some live debugging, realized our locking mechanism wasn't good enough either, because there is no guarantee in the ordering of when locks are released. So if three messages are sent in a row, the first will be sent first (assuming it doesn't lose a race in threading) then the 2nd and 3rd will block on the lock. Once the first is sent, the lock is released, but there is no guarantee the 2nd gets it, the third might and then you get more things out of order.

Noodled on this a while trying to come up with various options and in the end I think the following is the best solution:
1. our send_msg_task now supports a "task" that is a list of msgs to send. If it gets more than one msg, then it spaces them out by a second.
2. modified our flow processing to make sure to send all msgs for a set of actions / rulesets / steps at once for a single contact
3. change our msg queueing to batch up msgs by contact, essentially guaranteeing their serialized order

I've said it before, but I think this should finally resolve out of order messages on flows where more than one message is sent at once, this hopefully will address people who want a "pause" method in a flow. This does have SOME ramifications on sending performance, say in a giant broadcast with multiple messages in a row on a zillion contacts, but I think some of that is made up by having fewer tasks as well, so I'm willing to live with it until we see it as a problem.

### CODE REVIEW

File: temba/channels/tasks.py [hunk 1]
```diff
@@ -33,21 +38,32 @@ def send_msg_task():
...
+    msg_tasks = pop_task(SEND_MSG_TASK)
 
     # it is possible we have no message to send, if so, just return
-    if not cur_task:
+    if not msg_tasks:
         return
 
-    msg = dict_to_struct('MockMsg', cur_task,
-                         datetime_fields=['modified_on', 'sent_on', 'created_on', 'queued_on', 'next_attempt'])
+    if not isinstance(msg_tasks, list):
```
Comment (nicpottier) [inline temba/channels/tasks.py]: backwards compatibility so if we ship when we have single element tasks we don't blow up.

File: temba/channels/tasks.py [hunk 2]
```diff
@@ -33,21 +38,32 @@ def send_msg_task():
...
+        msg_tasks = [msg_tasks]
 
-    # send it off
     r = get_redis_connection()
 
-    # acquire a lock on our contact to make sure sending is always serialized
+    # acquire a lock on our contact to make sure two sets of msgs aren't being sent at the same time
```
Comment (nicpottier) [inline temba/channels/tasks.py]: Still good to have, this should be very uncommon to lock on though.

File: temba/flows/models.py [hunk 3]
```diff
@@ -565,7 +565,7 @@ def find_and_handle(cls, msg, started_flows=None, voice_response=None, triggered
 
     @classmethod
     def handle_destination(cls, destination, step, run, msg,
-                           started_flows=None, is_test_contact=False, user_input=False, triggered_start=False):
+                           started_flows=None, is_test_contact=False, user_input=False, triggered_start=False, trigger_send=True):
```
Comment (nicpottier) [inline temba/flows/models.py]: We hard ordering issues on new flow starts in that the first broadcast wasn't being sent until AFTER handle_destination was called. This now returns the messages created for sending by the caller in the case of flow starts.

File: temba/flows/tests.py [hunk 4]
```diff
@@ -4830,50 +4831,87 @@ def test_flow_batch_start(self):
...
         self.channel = Channel.create(self.org, self.user, 'KE', 'EX', None, '+250788123123', scheme='tel',
                                       config=dict(send_url='https://google.com'), uuid='00000000-0000-0000-0000-000000001234')
 
     def tearDown(self):
-        super(TwoInRowTest, self).tearDown()
+        super(OrderingTest, self).tearDown()
 
     def test_two_in_row(self):
-        flow = self.get_flow('two-in-row')
+        flow = self.get_flow('ordering')
```
Comment (nicpottier) [inline temba/flows/tests.py]: This is a super goofy flow that has 4 messages across 3 different action sets before waiting for input, then does the same thing again.

**Comment by ericnewcomer** [comment]
charmed i'm sure!

**Pull request merged**